src/java.base/share/classes/jdk/internal/module/ModuleInfo.java

```diff
@@ -405,14 +408,23 @@ private Builder readModuleAttribute(DataInput in, ConstantPool cpool, int major)
                     throw invalidModuleDescriptor("The requires entry for java.base"
                                                   + " has ACC_SYNTHETIC set");
                 }
+                // requires transitive java.base is illegal unless:
+                // - the major version is 53 (JDK 9), or:
+                // - the classfile is a preview classfile, or:
+                // - the module is deemed to be participating in preview
+                //   (i.e. the module is a java.* module)
+                // requires static java.base is illegal unless:
+                // - the major version is 53 (JDK 9), or:
                 if (major >= 54
-                    && (mods.contains(Requires.Modifier.TRANSITIVE)
+                    && ((mods.contains(Requires.Modifier.TRANSITIVE) &&
+                         !previewClassfile &&
+                         !mn.startsWith("java."))
```

Comment:
Just to double, will be JLS update specify that it's only allow for java.* aggregator modules?

Reply:
Only modules that participate in preview should be able to use `requires transitive java.base;` without having the preview flag set. And only modules in the JDK can participate in preview. I believe the JVMS will be updated to cover this.

We may want to restrict the condition here to only `java.se`, I am not sure. (Unless there are other modules that would use the directive, but I am not aware about any right now.)

Normally, the modules that "participate in preview" are those to which `java.base` exports the `jdk.internal.javac` package (and hence can use the `ParticipatesInPreview` annotation), but it didn't seem I can check the `java.base`'s exports at this place, so resorted to this hard-coded check. Hopefully, should only be temporary.